noir-projects/aztec-nr/aztec/src/oracle/notes.nr

```diff
@@ -200,17 +204,42 @@ pub unconstrained fn check_nullifier_exists(inner_nullifier: Field) -> bool {
 unconstrained fn check_nullifier_exists_oracle(_inner_nullifier: Field) -> Field {}
 
 /// Returns the tagging secret for a given sender and recipient pair, siloed for the current contract address.
+/// Includes the last known index used for tagging with this secret.
 /// For this to work, PXE must know the ivpsk_m of the sender.
 /// For the recipient's side, only the address is needed.
 pub unconstrained fn get_app_tagging_secret(
     sender: AztecAddress,
     recipient: AztecAddress,
-) -> Field {
-    get_app_tagging_secret_oracle(sender, recipient)
+) -> IndexedTaggingSecret {
+    let result = get_app_tagging_secret_oracle(sender, recipient);
+    IndexedTaggingSecret::deserialize(result)
 }
 
 #[oracle(getAppTaggingSecret)]
 unconstrained fn get_app_tagging_secret_oracle(
     _sender: AztecAddress,
     _recipient: AztecAddress,
-) -> Field {}
+) -> [Field; INDEXED_TAGGING_SECRET_LENGTH] {}
+
+/// Returns the tagging secrets for a given recipient and all the senders in PXE's address book,
+// siloed for the current contract address.
+/// Includes the last known index used for tagging with this secret.
+/// For this to work, PXE must know the ivsk_m of the recipient.
+pub unconstrained fn get_app_tagging_secrets_for_senders(
+    recipient: AztecAddress,
+) -> [IndexedTaggingSecret] {
+    let results = get_app_tagging_secrets_for_senders_oracle(recipient);
+    let mut indexed_tagging_secrets = &[];
+    for i in 0..results.len() {
+        if i % 2 != 0 {
+            continue;
+        }
+        indexed_tagging_secrets = indexed_tagging_secrets.push_back(
+            IndexedTaggingSecret::deserialize([results[i], results[i + 1]]),
+        );
```

Comment:
Bit nasty. Can we not receive structs from an oracle?

Reply:
Also we should `assert(results.len() % 2 == 0)`.

Reply:
This has been ultimately removed